File: test/driver/java/mil/tatrc/physiology/testing/SETestCase.java
package mil.tatrc.physiology.testing;

import java.util.*;

import com.kitware.physiology.cdm.TestReport.TestReportData.TestCaseData;

import mil.tatrc.physiology.datamodel.SEEqualOptions;
import mil.tatrc.physiology.datamodel.properties.SEScalarTime;

public class SETestCase 
{
	protected String         name;
	protected SEScalarTime   duration;
	protected List<String>   failures = new ArrayList<String>();

  protected SEEqualOptions eqOpts;
	
	protected SETestCase()
	{
		eqOpts = new SEEqualOptions();
		eqOpts.trackError(true);
		eqOpts.trackDifferences(true);
	}
	
	public void reset()
	{
		this.name=null;
		if(this.duration!=null)
			this.duration.invalidate();
		this.failures.clear();
	}
	
	public static void load(TestCaseData src, SETestCase dst)
  {
    dst.reset();
    dst.setName(src.getName());
    for(String failure : src.getFailureList())
    	dst.failures.add(failure);
  }
	
	public static TestCaseData unload(SETestCase src)
  {
		TestCaseData.Builder dst = TestCaseData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SETestCase src, TestCaseData.Builder dst)
  {
    if(src.hasName())
      dst.setName(src.name);
    if(src.hasDuration())
    	dst.setDuration(SEScalarTime.unload(src.duration));
    for(String failure : src.failures)
    	dst.getFailureList().add(failure);
  }
  
  public String  getName() { return this.name;}
  public void    setName(String name){this.name=name;}
  public boolean hasName(){return this.name==null?false:true;}
  
  public SEScalarTime getDuration() 
  { 
    if(this.duration==null)
      this.duration=new SEScalarTime();
    return this.duration;
  }
  public boolean      hasDuration() {return this.duration==null?false:this.duration.isValid();}
	
  public boolean hasFailures(){return failures.size()>1;}
  public void AddFailure(String f){failures.add(f);}
  public List<String> getFailures(){ return Collections.unmodifiableList(this.failures); }
  
  SEEqualOptions getEqualOptions(){return eqOpts;}
}

File: cdm/cpp/utils/testing/SETestCase.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "utils/testing/SETestCase.h"
#include "properties/SEScalarTime.h"

SETestCase::SETestCase(Logger* logger) : Loggable(logger)
{
  SETestCase("default", logger);
}
SETestCase::SETestCase(const std::string& name, Logger* logger) : Loggable(logger)
{
  m_Name = name;
}

SETestCase::~SETestCase()
{
  Clear();
}

void SETestCase::Clear()
{
  m_Failure.clear();
  m_Duration.SetValue(0, TimeUnit::s);
  DELETE_VECTOR(m_CaseEqualsErrors);
}

void SETestCase::Reset()
{
  m_Failure.clear();
  m_Duration.SetValue(0,TimeUnit::s);
  DELETE_VECTOR(m_CaseEqualsErrors);
}

void SETestCase::Load(const cdm::TestReportData_TestCaseData& src, SETestCase& dst)
{
  SETestCase::Serialize(src, dst);
}
void SETestCase::Serialize(const cdm::TestReportData_TestCaseData& src, SETestCase& dst)
{
  dst.Clear();

  dst.m_Name = src.name();
  if(src.has_duration())
    SEScalarTime::Load(src.duration(), dst.GetDuration());

  SETestErrorStatistics* ex;
  for (int i = 0; i<src.errorstats_size(); i++)
  {
    auto eData = src.errorstats(i);
    ex = new SETestErrorStatistics(dst.GetLogger());
    SETestErrorStatistics::Load(eData, *ex);
    dst.m_CaseEqualsErrors.push_back(ex);
  }

  for (int i=0; i<src.failure_size(); i++)
    dst.m_Failure.push_back(src.failure(i));
}

cdm::TestReportData_TestCaseData* SETestCase::Unload(const SETestCase& src)
{
  cdm::TestReportData_TestCaseData* dst = new cdm::TestReportData_TestCaseData();
  SETestCase::Serialize(src,*dst);
  return dst;
}
void SETestCase::Serialize(const SETestCase& src, cdm::TestReportData_TestCaseData& dst)
{
  dst.set_name(src.m_Name);

  dst.set_allocated_duration(SEScalarTime::Unload(src.m_Duration));

  for (std::string s : src.m_Failure)
    dst.mutable_failure()->Add(s.c_str());

  for (auto s : src.m_CaseEqualsErrors)
    dst.mutable_errorstats()->AddAllocated(SETestErrorStatistics::Unload(*s));
}

void SETestCase::SetName(const std::string& Name)
{
  m_Name = Name;
}

std::string SETestCase::GetName() const
{
  return m_Name;
}

SEScalarTime& SETestCase::GetDuration()
{
  return m_Duration;
}


void SETestCase::AddFailure(std::stringstream &err)
{
  AddFailure(err.str());
  err.str("");
  err.clear();
}
void SETestCase::AddFailure(const std::string& err)
{
  m_Failure.push_back(err);
  Error(err);
}
const std::vector<std::string>& SETestCase::GetFailures()
{
  return m_Failure;
}

SETestErrorStatistics& SETestCase::CreateErrorStatistic()
{
  SETestErrorStatistics* err = new SETestErrorStatistics(GetLogger());
  m_CaseEqualsErrors.push_back(err);
  return *err;
}
const std::vector<SETestErrorStatistics*>& SETestCase::GetErrorStatistics() const
{
  return m_CaseEqualsErrors;
}
